File: 29. Divide Two Integers/solution.js
/**
 * @param {number} dividend
 * @param {number} divisor
 * @return {number}
 */
var divide = function(dividend, divisor) {
  var result = 0;
  var step, incre;
  var negative = (dividend < 0) ^ (divisor < 0);

  if (divisor == 0) return 0;

  dividend = Math.abs(dividend);
  divisor = Math.abs(divisor);

  while (dividend > divisor) {
    step = divisor;
    incre = 1;
    while (dividend > step) {
      dividend -= step;
      result += incre;
      step <<= 1;
      incre <<= 1;
    }
  }
  if (negative) {
    return Math.min(2147483647, result);
  } else {
    return Math.max(-2147483648, result);
  }

};

File: 29. Divide Two Integers/solution.java
public class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == Integer.MIN_VALUE && divisor == -1) return Integer.MAX_VALUE;
        if (dividend > 0 && divisor > 0) return divideHelper(-dividend, -divisor);
        else if (dividend > 0) return -divideHelper(-dividend,divisor);
        else if (divisor > 0) return -divideHelper(dividend,-divisor);
        else return divideHelper(dividend, divisor);
    }

    private int divideHelper(int dividend, int divisor) {
        if (divisor < dividend) return 0;
        int divideResult = 0;
        int result = 1;
        int multiDivisor = divisor;
        int shiftCount = 0;
        while (dividend < multiDivisor && (multiDivisor << 1) < 0 && shiftCount < 30) {
            multiDivisor <<= 1;
            result <<= 1;
            shiftCount ++;
        }
        while (dividend <= divisor) {
            while (dividend > multiDivisor) {
                multiDivisor >>= 1;
                result >>= 1;
            }
            divideResult += result;
            dividend -= multiDivisor;
        }
        return divideResult;
    }
}